records/2024-04-05.database-activity-logging.md

```diff
@@ -0,0 +1,74 @@
+# Database activity logging
+
+Date: 2024-04-05
+
+## Status
+
+**DRAFT**
+
+## Context
+
+To improve GC Notify's security posture, database activity logging should be enabled.  This will give the product and security teams visibility into the database activity.  It will also allow for the creation of automated alerting rules to detect malicious or unusual behaviour.  
+
+This work is related to the most recent Security Assessment and Authorization (SA&A) performed on Notify, which identified the [AU-6 - Audit Review, Analysis, and Reporting](https://www.cyber.gc.ca/en/guidance/annex-3a-security-control-catalogue-itsg-33#a33au6) control as required.
+
+## Options
+
+There are two options for auditing database activity with an RDS Aurora PostgreSQL database cluster:
+
+1. [database activity streams](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/DBActivityStreams.Overview.html); and
+2. the [PostgreSQL audit extension (pgAudit)](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/Appendix.PostgreSQL.CommonDBATasks.pgaudit.html).
+
+### Option 1: database activity streams
+
+A database activity stream logs all queries performed by the databases that are part of an Aurora cluster. The activity stream is encrypted with a KMS Customer Managed Key (CMK) and can be integrated with other AWS services.  No database instance configuration or storage space is needed to setup and use the activity stream.
+
+The proposed solution for GC Notify would be to create an asynchronous activity stream to minimize database performance impact.  This stream would then be ingested by a Kinesis Firehose which would:
+
+1. decrypt the activity events,
+1. filter events that should not be captured; and 
+1. write activity logs to an S3 bucket for auditing and alerting.
+
+![Database activity stream architecture](./diagrams/2024-04-05.database-activity-logging/activity-stream-arch.png)
+_Figure 1: [Database activity stream architecture](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/DBActivityStreams.Overview.html#DBActivityStreams.Overview.how-they-work)_
+
+In order to respect Notify's privacy policy, the S3 activity logs would be deleted after 7 days.
```

Comment:
Could we redirect s3 logs into different buckets, with the idea that the app log is kept 7 days but the admin users are kept 1+ year?

Reply:
Absolutely - I think that is a very good idea.   Once we've got IAM database authentication setup, it will be easy to separate app queries from Notify team member queries and apply different retention policies on them.

I'd also like to setup Sentinel forwarding so that we'd be able to create alerts for unusual database activity.  

I'll update the ADR with this details!